File: sdks/cpp/connections/gRPC/src/ExecuteCommand.cpp
/*
 * Copyright 2024 Ross Video Ltd
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS”
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

// common
#include <Tags.h>

// connections/gRPC
#include <ExecuteCommand.h>

// type aliases
using catena::common::ParamTag;
using catena::common::Path;

#include <iostream>
#include <thread>
#include <fstream>
#include <vector>
#include <iterator>
#include <filesystem>

//Counter for generating unique object IDs - static, so initializes at start
int CatenaServiceImpl::ExecuteCommand::objectCounter_ = 0;

/** 
 * Constructor which initializes and registers the current ExecuteCommand
 * object, then starts the process
 */
CatenaServiceImpl::ExecuteCommand::ExecuteCommand(CatenaServiceImpl *service, Device &dm, bool ok)
    : service_{service}, dm_{dm}, stream_(&context_),
        status_{ok ? CallStatus::kCreate : CallStatus::kFinish} {
    service->registerItem(this);
    objectId_ = objectCounter_++;
    proceed(service, ok);  // Start the process
}

/**
 * Manages gRPC command execution process by transitioning between states and
 * handling errors and responses accordingly 
 */
void CatenaServiceImpl::ExecuteCommand::proceed(CatenaServiceImpl *service, bool ok) {
    std::cout << "ExecuteCommand proceed[" << objectId_ << "]: " << timeNow()
                << " status: " << static_cast<int>(status_) << ", ok: " << std::boolalpha << ok
                << std::endl;

    // If the process is cancelled, finish the process
    if(!ok){
        std::cout << "ExecuteCommand[" << objectId_ << "] cancelled\n";
        status_ = CallStatus::kFinish;
    }

    //State machine to manage the process
    switch (status_) {
        /**
         * Initial state, sets up reques to execute command and transitions to
         * kProcess
         */
        case CallStatus::kCreate:
            status_ = CallStatus::kProcess;
            service_->RequestExecuteCommand(&context_, &stream_, service_->cq_, service_->cq_, this);
            break;

        /**
         * Processes the command by reading the initial request from the client
         * and transitioning to kRead
         */
        case CallStatus::kProcess:
            new ExecuteCommand(service_, dm_, ok);
            /** 
             * ExecuteCommand RPC always begins with reading initial request
             * from client
             */
            status_ = CallStatus::kRead;
            stream_.Read(&req_, this);
            break;

        /** 
         * Reads the command request from the client and transitions to kRead.
         * Should always tranistion to this state after calling stream_.Read()
         */
        case CallStatus::kRead:
        {
            catena::exception_with_status rc{"", catena::StatusCode::OK};
            std::unique_ptr<IParam> command = nullptr;
            // Check if authorization is enabled
            if (service_->authorizationEnabled()) {
                std::vector<std::string> clientScopes = service->getScopes(context_);
                catena::common::Authorizer authz{clientScopes};
                command = dm_.getCommand(req_.oid(), rc, authz);
            } else {
                command = dm_.getCommand(req_.oid(), rc, catena::common::Authorizer::kAuthzDisabled);
            }

            // If the command is not found, return an error
            if (command == nullptr) {
                status_ = CallStatus::kFinish;
                stream_.Finish(Status(static_cast<grpc::StatusCode>(rc.status), rc.what()), this);
                break;
            }
            // Execute the command
            res_ = command->executeCommand(req_.value());
            /**
             * @todo: Implement streaming response
             * For now we are not streaming the response so we can finish the
             * call
             */
            status_ = CallStatus::kPostWrite; 
            stream_.Write(res_, this);
            break;
        }

        /**
         * Status after reading the request, which transitions to kRead to
         * handle subsequent requests
         */
        case CallStatus::kWrite:
            status_ = CallStatus::kRead;
            stream_.Read(&req_, this);
            break;

        // Status after finishing writing the response, transitions to kFinish
        case CallStatus::kPostWrite:
            status_ = CallStatus::kFinish;
            stream_.Finish(Status::OK, this);
            break;

        /**
         * Deregisters the current ExecuteCommand object and finishes the
         * process
         */
        case CallStatus::kFinish:
            std::cout << "ExecuteCommand[" << objectId_ << "] finished\n";
            service->deregisterItem(this);
            break;

        // Throws an error if the state is not recognized
        default:
            status_ = CallStatus::kFinish;
            grpc::Status errorStatus(grpc::StatusCode::INTERNAL, "illegal state");
            stream_.Finish(errorStatus, this);
    }
}

File: sdks/go/examples/executeCommand/executecommand.go
/*
 * Copyright 2026 Ross Video Ltd
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * @brief Shared ExecuteCommand example logic for REST and gRPC.
 * @file executecommand.go
 * @copyright Copyright © 2026 Ross Video Ltd
 * @author Nelson Daniels (nelson.daniels@rossvideo.com)
 */

package executecommand

import (
	"fmt"
	"sync"
	"time"

	"github.com/rossvideo/catena/sdks/go/examples/exampleutil"
	"github.com/rossvideo/catena/sdks/go/pkg/catena"
	"github.com/rossvideo/catena/sdks/go/pkg/logger"
)

type CommandHandler func(payload any) (catena.CommandResult, catena.StatusResult)

type CounterState struct {
	mu      sync.RWMutex
	value   int32
	running bool
}

func (c *CounterState) GetValue() int32 {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.value
}

func (c *CounterState) IsRunning() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.running
}

func (c *CounterState) Start() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.running = true
}

func (c *CounterState) Stop() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.running = false
}

func (c *CounterState) Add(n int32) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value += n
}

func (c *CounterState) Reset() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value = 0
}

func (c *CounterState) Increment() {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.running {
		c.value++
	}
}

var SlotList = []uint16{0}

func BuildResponse(counter *CounterState) map[string]any {
	var runningVal int32 = 0
	if counter.IsRunning() {
		runningVal = 1
	}
	return map[string]any{
		"counter": int32(counter.GetValue()),
		"running": runningVal,
	}
}

// RegisterHandlers registers the GetValue and ExecuteCommand handlers for the
// counter demo. It also starts a background goroutine that increments the
// counter every second while running.
func RegisterHandlers(srv catena.Server) {
	counter := &CounterState{}

	go func() {
		ticker := time.NewTicker(1 * time.Second)
		defer ticker.Stop()
		for range ticker.C {
			if counter.IsRunning() {
				counter.Increment()
				logger.Info("Counter tick", "value", counter.GetValue())
			}
		}
	}()

	commands := map[string]CommandHandler{
		"start": func(payload any) (catena.CommandResult, catena.StatusResult) {
			if counter.IsRunning() {
				logger.Info("Start command - already running")
				val, _ := catena.ToValue(BuildResponse(counter))
				return catena.CommandReply(val)
			}
			counter.Start()
			logger.Info("Counter started", "value", counter.GetValue())
			val, _ := catena.ToValue(BuildResponse(counter))
			return catena.CommandReply(val)
		},

		"stop": func(payload any) (catena.CommandResult, catena.StatusResult) {
			if !counter.IsRunning() {
				logger.Info("Stop command - already stopped")
				val, _ := catena.ToValue(BuildResponse(counter))
				return catena.CommandReply(val)
			}
			counter.Stop()
			logger.Info("Counter stopped", "value", counter.GetValue())
			val, _ := catena.ToValue(BuildResponse(counter))
			return catena.CommandReply(val)
		},

		"add10": func(payload any) (catena.CommandResult, catena.StatusResult) {
			counter.Add(10)
			logger.Info("Added 10 to counter", "value", counter.GetValue())
			val, _ := catena.ToValue(BuildResponse(counter))
			return catena.CommandReply(val)
		},

		"reset": func(payload any) (catena.CommandResult, catena.StatusResult) {
			counter.Reset()
			logger.Info("Counter reset")
			return catena.CommandNoResponse()
		},

		"error": func(payload any) (catena.CommandResult, catena.StatusResult) {
			logger.Info("Error command executed")
			return catena.CommandExceptionResult(
				"DemoError",
				"This is a demonstration of a command exception",
				catena.NewPolyglotText("en", "An example error occurred"),
			)
		},
	}

	srv.RegisterGetValueHandler(0, func(slot uint16, fqoid string) (catena.Value, catena.StatusResult) {
		if fqoid == "counter" {
			val, _ := catena.ToValue(BuildResponse(counter))
			return catena.Reply(val)
		}
		return catena.ReplyError[catena.Value](catena.NOT_FOUND, "parameter not found: "+fqoid)
	})

	srv.RegisterExecuteCommandHandler(0, func(slot uint16, commandFqoid string, payload any) (catena.CommandResult, catena.StatusResult) {
		logger.Info("ExecuteCommand", "slot", slot, "command", commandFqoid)

		handler, ok := commands[commandFqoid]
		if !ok {
			logger.Warning("Command not found", "slot", slot, "command", commandFqoid)
			return catena.CommandError(catena.NOT_FOUND, "Command not found: "+commandFqoid)
		}

		return handler(payload)
	})
}

// RunExample encapsulates the full example lifecycle:
// SDK init, signal handling, server creation, handler registration, and graceful shutdown.
func RunExample(appName string) {
	exampleutil.RunExample(exampleutil.RunConfig{
		AppName:          appName,
		Slots:            SlotList,
		RegisterHandlers: RegisterHandlers,
		OnReady: func(cfg catena.Config) {
			port := cfg.Port
			logger.Info("=======================================================")
			logger.Info("ExecuteCommand Example", "transport", appName)
			logger.Info("=======================================================")
			logger.Info("Listening", "port", port)
			logger.Info("")
			logger.Info("Web UI available at:")
			logger.Info(fmt.Sprintf("  http://localhost:%d/", port))
			logger.Info("")
			logger.Info("=======================================================")
		},
	})
}
